chips/lowrisc/src/uart.rs

```diff
@@ -192,18 +192,26 @@ impl<'a> Uart<'a> {
 
             // Read from the transmit buffer and send bytes to the UART hardware
             // until either the buffer is empty or the UART hardware is full.
-            self.tx_buffer.map(|tx_buf| {
+            if self.tx_buffer.map_or(false, |tx_buf| {
                 let tx_len = len - idx;
 
                 for i in 0..tx_len {
                     if regs.status.is_set(status::txfull) {
-                        break;
+                        return false;
                     }
                     let tx_idx = idx + i;
                     regs.wdata.write(wdata::data.val(tx_buf[tx_idx] as u32));
                     self.tx_index.set(tx_idx + 1)
                 }
-            });
+                true
+            }) {
+                // We sent everything to the UART hardware
+                self.tx_client.map(|client| {
+                    self.tx_buffer.take().map(|tx_buf| {
+                        client.transmitted_buffer(tx_buf, self.tx_len.get(), Ok(()));
```

Comment:
Unfortunately we can't call a callback here. If I'm reading this driver correctly, `transmit_buffer()` is called from hil::uart::transmit::transmit_buffer(), which means a callback could occur in response to a HIL command. Per Tock's callback model, callbacks can _only_ be triggered in response to interrupts.

Reply:
Callbacks do not have to be triggered in response to interrupts; they just can't be invoked on the same call stack from the causing downcall. This generally means that it needs to either be from an interrupt or a deferred procedure call.

But yes, invoking `transmitted_buffer` here is incorrect.